src/main/java/io/vertx/core/http/impl/SharedHttpClient.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) 2011-2021 Contributors to the Eclipse Foundation
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ */
+
+package io.vertx.core.http.impl;
+
+import io.vertx.core.*;
+import io.vertx.core.http.HttpClient;
+import io.vertx.core.http.HttpClientOptions;
+import io.vertx.core.impl.VertxInternal;
+import io.vertx.core.shareddata.LocalMap;
+
+public class SharedHttpClient extends WrappedHttpClient implements Closeable {
+
+  private static final String MAP_NAME = "__vertx.shared.httpClients";
+  public static final String DEFAULT_CLIENT_NAME = "SharedHttpClient.DEFAULT";
+
+  private final VertxInternal vertx;
+  private final String name;
+
+  private SharedHttpClient(VertxInternal vertx, String name, HttpClient delegate) {
+    super(delegate);
+    this.vertx = vertx;
+    this.name = name;
+  }
+
+  public static SharedHttpClient create(VertxInternal vertx, String name, HttpClientOptions options) {
+    LocalMap<String, HttpClientHolder> localMap = vertx.sharedData().getLocalMap(MAP_NAME);
+    HttpClient client;
+    HttpClientHolder current, candidate;
+    for (; ; ) {
+      current = localMap.get(name);
+      if (current != null) {
+        candidate = current.increment();
+        if (candidate != null && localMap.replaceIfPresent(name, current, candidate)) {
+          client = candidate.get();
+          break;
+        }
+      } else {
+        candidate = new HttpClientHolder();
+        if (localMap.putIfAbsent(name, candidate) == null) {
+          candidate = candidate.init(vertx, options);
+          client = candidate.get();
+          localMap.put(name, candidate);
+          break;
+        }
+      }
+    }
+    return new SharedHttpClient(vertx, name, client);
+  }
+
+  @Override
+  public void close(Handler<AsyncResult<Void>> handler) {
+    Future<Void> future = close();
+    if (handler != null) {
+      future.onComplete(handler);
+    }
+  }
+
+  @Override
+  public Future<Void> close() {
+    Promise<Void> promise = vertx.promise();
+    LocalMap<String, HttpClientHolder> localMap = vertx.sharedData().getLocalMap(MAP_NAME);
+    HttpClientHolder current, candidate;
+    for (; ; ) {
+      current = localMap.get(name);
+      candidate = current.decrement();
```

Comment:
we should be using map compute instead of removeIfPresent / replace

Reply:
:+1:

Reply:
Looking again, I did not use compute because we need a way to determine when to actually close the underlying client.

Also, I think using remove/replace is not too critical in this case, considering the low concurrency and frequency of invocations.